contrib/agingBloom.c

```diff
@@ -94,315 +102,446 @@ static void destroySlice(blmSlice *slice) {
+                case 1: k = 15; l = 15; break;
+                case 2: k = 16; l = 22; break;
+                case 3: k = 17; l = 36; break;
+                case 4: k = 18; l = 63; break;
+                case 5: k = 19; l = 117; break;
+                default: break;
+            }
+            break;
+        case 5:
+            switch (level) {
+                case 0: k = 17; l = 13; break;
+                case 1: k = 18; l = 16; break;
+                case 2: k = 19; l = 22; break;
+                case 3: k = 20; l = 33; break;
+                case 4: k = 21; l = 54; break;
+                case 5: k = 22; l = 95; break;
+                default: break;
+            }
+        default: break;
     }
-  break;
-  case 2:
-    switch (level) {
-      case 0: k = 7; l = 5; break;
-      case 1: k = 8; l = 8; break;
-      case 2: k = 9; l = 14; break;
-      case 3: k = 10; l = 25; break;
-      case 4: k = 11; l = 46; break;
-      case 5: k = 12; l = 88; break;
-      default: break;
-    }
-  break;
-  case 3:
-  switch (level) {
-    case 0: k = 10; l = 7; break;
-    case 1: k = 11; l = 9; break;
-    case 2: k = 12; l = 14; break;
-    case 3: k = 13; l = 23; break;
-    case 4: k = 14; l = 40; break;
-    case 5: k = 15; l = 74; break;
-    default: break;
-  }
-  break;
-  case 4:
-  switch (level) {
-    case 0: k = 14; l = 11; break;
-    case 1: k = 15; l = 15; break;
-    case 2: k = 16; l = 22; break;
-    case 3: k = 17; l = 36; break;
-    case 4: k = 18; l = 63; break;
-    case 5: k = 19; l = 117; break;
-    default: break;
-  }
-  break;
-  case 5:
-  switch (level) {
-    case 0: k = 17; l = 13; break;
-    case 1: k = 18; l = 16; break;
-    case 2: k = 19; l = 22; break;
-    case 3: k = 20; l = 33; break;
-    case 4: k = 21; l = 54; break;
-    case 5: k = 22; l = 95; break;
-    default: break;
-  }
-  default: break;
-  }
 
-  assert(capacity > l); // To avoid a mod 0 on the shift code
-  uint64_t sliceSize = (capacity * k) / (l * log(2));
+    assert(capacity > l); // To avoid a mod 0 on the shift code
+    uint64_t sliceSize = (capacity * k) / (l * log(2));
 
-  ageBloom_t* bf = APBF_createLowLevelAPI(k, l, ceil(sliceSize));
-  bf->capacity = capacity;
-  bf->errorRate = error;
-  return bf;
+    ageBloom_t* bf = APBF_createLowLevelAPI(k, l, ceil(sliceSize));
+    bf->capacity = capacity;
+    bf->errorRate = error;
+    return bf;
 }
 
-ageBloom_t *APBF_createTimeAPI(int error, uint64_t capacity, timestamp_t frequency) {
-  int8_t level = 4;
-  ageBloom_t *apbf = APBF_createHighLevelAPI(error, capacity, level);
-  apbf->assessFreq = frequency;
-  return apbf;
+// Used for time based APBF
+ageBloom_t *APBF_createTimeAPI(int error, uint64_t capacity, int8_t level, timestamp_t frequency) {
+    ageBloom_t *apbf = APBF_createHighLevelAPI(error, capacity, level);
+    assert(apbf);
+    ts = apbf->assessFreq = frequency;
+    apbf->slices[apbf->numHash].timestamp = ts;
```

Comment:
What are we doing here?

Reply:
I changed some of the names as you suggested in our last meeting: ts is now currTime and assessFreq is timeSpan, but we can change them again if they're not good.
Initially, I set the currTime to be equal to the timeSpan, so when we're doing the verification whether a slice is in the specified time span (slices[i].timestamp < **apbf->currTime - apbf->timeSpan**) we don't get a negative value. The same goes for the second line of code when the time between shifts is calculated in the function predictSize.
In alternative, I could also set the currTime initially to 0 and, every time the verification is made, simply add to it the value of timeSpan, like this: slices[i].timestamp < (apbf->currTime + timeSpan) - timeSpan. Maybe the second option is better because the variable currTime doesn't grow so fast, since it starts as 0 instead of a large number. But there's also the downside of having to do this add operation in more than one place every time there's a shift. What do you think?